plugins/onion.js

```diff
@@ -1,42 +1,82 @@
-var net
-try {
-  net = require('net')
-} catch (_) {}
-
-var proxy = require('tproxy')
+var socks = require('socks');
 var toPull = require('stream-to-pull-stream')
 
+var options = {
+    proxy: {
+        ipaddress: "localhost",
+        port: 9050, // default tor port
+        type: 5
+    },
+}
+
 module.exports = function (opts) {
-  opts.allowHalfOpen = opts.allowHalfOpen !== false
   return {
     name: 'onion',
-    server: function (onConnection) {
-      var server = net.createServer(opts, function (stream) {
-        var proxyStream = proxy(stream, { port: 9050 }) // tor default
-        onConnection(proxyStream = toPull.duplex(proxyStream))
-      }).listen(opts)
-      return function () {
-        server.close()
-      }
+      server: function (onConnection) {
+          var serverOpts = {
+              proxy: options.proxy,
+              command: "bind",
+              target: {
+                  host: opts.host,
+                  port: opts.port
+              }
+          }
+          var controlSocket
+          socks.createConnection(serverOpts, function (err, socket) {
+              controlSocket = socket
+
+              socket.on('data', function(data) {
+                  onConnection(data = toPull.duplex(data))
+              })
+
+              // Remember to resume the socket stream.
+              socket.resume()
+          })
+          return function () {
+              controlSocket.end()
```

Comment:
race condition here. need to check if `controlSocket` is set yet,
and if not, set a `closed` var, and in the callback `createConnection` close the stream if closed is true.